src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+    ) -> F::Handle {
+        match file {
+            Either::Left(f) => {
+                // TODO: should we handle the second disk io error
+                block_on(f).unwrap().unwrap().unwrap()
+            }
+            Either::Right(fd) => fd,
+        }
+    }
+
+    fn process(fd: &Arc<F::Handle>, task: FileTask) -> IoResult<TaskRes<F>> {
+        match task {
+            FileTask::Truncate(offset) => fd.truncate(offset).map(TaskRes::Truncate),
+            FileTask::FileSize => fd.file_size().map(|s| TaskRes::FileSize(s)),
+            FileTask::Sync => fd.sync().map(TaskRes::Sync),
+            FileTask::Write { offset, bytes } => fd.write(offset, &bytes).map(|s| TaskRes::Write(s)),
+            FileTask::Allocate { offset, size } => fd.allocate(offset, size).map(TaskRes::Allocate),
+            FileTask::NewWriter { fs } => fs.new_writer(fd.clone()).map(|w| TaskRes::NewWriter(w)),
+            FileTask::Stop => unreachable!(),
+        }
+    }
+
+    fn new_writer(&self) -> IoResult<HedgedWriter<F>> {
+        let task1 = FileTask::NewWriter(self.base.clone());
+        let task2 = FileTask::NewWriter(self.base.clone());
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.disk1.send((task1, cb1)).unwrap();
+        self.disk2.send((task2, cb2)).unwrap();
+
+        select! {
+            res1 = f1 => res1.unwrap().map(|w| HedgedWriter::new(self.disk1.clone(), self.disk2.clone(), 
+                Either::Right(w.unwrap()), Either::Left(f2) )),
+            res2 = f2 => res2.unwrap().map(|w| HedgedHandle::new(
+                Either::Left(f1), Either::Right(w.unwrap()) )),
+        }
+        select! {
+            res1 = f1 => res1.unwrap().map(|_| ()),
+            res2 = f2 => res2.unwrap().map(|_| ()),
+        }
+    }
+
+    fn read(&self, offset: usize, buf: &mut [u8]) -> IoResult<usize> {
+        // TODO: read simultaneously from both disks
+        // choose latest to perform read
+        let count1 = self.counter1.load(Ordering::Relaxed);
+        let count2 = self.counter2.load(Ordering::Relaxed);
+        match count1.cmp(&count2) {
+            std::cmp::Ordering::Equal => {
+                if let Some(fd) = self.fd1.read().unwrap().as_ref() {
+                    fd.read(offset, buf)
+                } else if let Some(fd) = self.fd2.read().unwrap().as_ref() {
+                    fd.read(offset, buf)
+                } else {
+                    panic!("Both fd1 and fd2 are None");
+                }
+            }
+            std::cmp::Ordering::Greater => {
+                self.fd1.read().unwrap().as_ref().unwrap().read(offset, buf)
+            }
+            std::cmp::Ordering::Less => {
+                self.fd2.read().unwrap().as_ref().unwrap().read(offset, buf)
+            }
+        }
+    }
+
+    fn write(&self, offset: usize, content: &[u8]) -> IoResult<usize> {
```

Comment:
Is that possible that when we do a write, there is an another thread that's doing read?

Reply:
yes

Reply:
Since based on the design, we achieve an eventual consistency on the two disks.  Meaning, at a time,  the content of the two disks can be different.   If the writer is writing disk 1,  while the reader read from disk 2,  looks like the raft engine would make a wrong action based on the wrong point of view.

Reply:
Note the offset is specified. The reader always reads from the newer disk at the point of time. If the reader can read from disk2, it indicates that disk2 must have the data of that offset. It doesn't need to care about the latest write because it's not reading the latest writes' offset